ucp/_libs/ucx_api.pyx

```diff
@@ -0,0 +1,490 @@
+    which should freed using `ucp_config_release()`.
+    """
+    cdef ucp_config_t *config
+    cdef ucs_status_t status
+    status = ucp_config_read(NULL, NULL, &config)
+    if status != UCS_OK:
+        raise UCXConfigError(
+            "Couldn't read the UCX options: %s" %
+            ucs_status_string(status).decode("utf-8")
+        )
+
+    # Modify the UCX configuration options based on `config_dict`
+    for k, v in user_options.items():
+        status = ucp_config_modify(config, k.encode(), v.encode())
+        if status == UCS_ERR_NO_ELEM:
+            raise UCXConfigError("Option %s doesn't exist" % k)
+        elif status != UCS_OK:
+            msg = "Couldn't set option %s to %s: %s" % \
+                  (k, v, ucs_status_string(status).decode("utf-8"))
+            raise UCXConfigError(msg)
+    return config
+
+
+cdef ucx_config_to_dict(ucp_config_t *config):
+    """Returns a dict of a UCX config"""
+    cdef char *text
+    cdef size_t text_len
+    cdef FILE *text_fd = open_memstream(&text, &text_len)
+    assert(text_fd != NULL)
+    ret = {}
+    ucp_config_print(config, text_fd, NULL, UCS_CONFIG_PRINT_CONFIG)
+    fflush(text_fd)
+    cdef unicode py_text = text.decode()
+    for line in py_text.splitlines():
+        k, v = line.split("=")
+        k = k[len("UCX_"):]
+        ret[k] = v
+    fclose(text_fd)
+    free(text)
+    return ret
+
+
+def get_current_options():
+    """
+    Returns the current UCX options
+    if UCX were to be initialized now.
+    """
+    cdef ucp_config_t *config = _read_ucx_config({})
+    ret = ucx_config_to_dict(config)
+    ucp_config_release(config)
+    return ret
+
+
+def get_ucx_version():
+    cdef unsigned int a, b, c
+    ucp_get_version(&a, &b, &c)
+    return (a, b, c)
+
+
+cdef void _listener_callback(ucp_ep_h ep, void *args) except *:
+    cdef dict cb_data = <dict> args
+    cb_data['cb_func'](
+        ucx_ep_create(ep),
+        *cb_data['cb_args']
+    )
+
+
+cdef class UCXListener:
+    cdef:
+        ucp_listener_h handle
+        uint16_t _port
+        dict cb_data
+        bint initialized
+
+    def __cinit__(
+        self,
+        UCXWorker worker,
+        uint16_t port,
+        cb_func,
+        cb_args
+    ):
+
+        self.initialized = False
+        self._port = port
+        self.cb_data = {
+            'cb_func': cb_func,
+            'cb_args': cb_args,
+        }
+
+        cdef ucp_listener_params_t params
+        cdef ucp_listener_accept_callback_t _listener_cb = (
+            <ucp_listener_accept_callback_t>_listener_callback
+        )
+        if c_util_get_ucp_listener_params(&params,
+                                          port,
+                                          _listener_cb,
+                                          <void*> self.cb_data):
+            raise MemoryError("Failed allocation of ucp_ep_params_t")
+
+        logging.info("create_listener() - Start listening on port %d" % port)
```

Comment:
Have we setup some kind of logging handler somewhere?

Reply:
Yes here: https://github.com/rapidsai/ucx-py/blob/branch-0.13/ucp/__init__.py#L15
We need to clean up the logging and follow best practice, which i honestly don't know what is :)
But let's do that in a new PR.

Reply:
Ok. Sounds good.